### ISSUE TITLE
How to remove a table from the schema before calling replace_schema?

### ISSUE THREAD

**Issue by ElixirMike** [hasura/graphql-engine#4718]
I'm working on an app where I will need to make changes to the schema, and then run a replace_schema API call with the new data.

I am using the export_schema api method to get the schema. I identify the table I want to delete/remove and then use a delete statement to remove it from the JSON object. If I inspect the json object, it looks totally fine, minus the table I removed.

delete freshMeta.data.tables[fieldTableIndex];

When i call the replace_schema method with the updated schema, I am receiving this error:

{"path":"$[1].tables[114]","error":"expecting an Object for TableMetadata","code":"parse-failed"}

Is there some other step I need to do in order to remove a specific table node from the schema before updating?

**Comment by tirumaraiselvan** [comment]
It appears your metadata structure is not exactly correct, can you paste the top-level structure of your metadata after the programmatic manipulation?

**Comment by ElixirMike** [comment]
Here is a stringified version of the resulting json
[schema_json.txt](https://github.com/hasura/graphql-engine/files/4609795/schema_json.txt)

**Comment by ElixirMike** [comment]
Also, here is the process I'm using ..

{
      "type" : "export_metadata",
                     "args": {}
 }

I get the resulting data in a variable freshMetaData.

I then run this to delete a table from the schema where fieldTableIndex is the index I want to remove.
delete freshMeta.data.tables[fieldTableIndex]; 

I then run 
{
                    "type" : "replace_metadata",
                     "args": ${JSON.stringify(freshMeta.data)}
                }

**Comment by tirumaraiselvan** [comment]
@ElixirMike I don't think you need to send the stringified version of the new metadata. 

Try this:

```
{
"type" : "replace_metadata",
"args": freshMeta.data
}
```

**Comment by ElixirMike** [comment]
yea, that was what I initially tried. I  get this error:
{"path":"$","error":"Error in $: Failed reading: not a valid json value","code":"invalid-json"}

when I used Stringify, it started working...but only for when I was appending new tables to schema. not removing

**Comment by tirumaraiselvan** [comment]
You should inspect and see why 

```
{
"type" : "replace_metadata",
"args": freshMeta.data
}
```

is not a valid json as reported by the server: `{"path":"$","error":"Error in $: Failed reading: not a valid json value","code":"invalid-json"}`

**Comment by ElixirMike** [comment]
Well, this happens even if I make no changes to the metadata.
So, if I simply Export the metaData, and then use that to run replace_metadata, with no changes, I get the error above. If I use stringify, it goes through.

**Comment by ElixirMike** [comment]
Can you confirm if you export metadata, and then try re-importing using the same, it gives error?
I am using React/Javascript . If you could provide a simply working example of how you
1. Export the metadata
2. Delete/remove a table from this
3. Replace_metadata

That would go a long way in helping me get through this issue. 

{
"type" : "export_metadata",
"args": {}
}

{
"type" : "replace_metadata",
"args": freshMeta.data
}

**Comment by tirumaraiselvan** [comment]
Hi @ElixirMike 

This is what I did:

1. I exported the metadata
2. I opened the exported file and enclosed the json in `{ "type": "replace_metadata", "args": <file-contents> }` and saved it as `replace_metadata.json`
3. I then curled the file: `curl -d "@replace_metadata.json" http://my-hasura-url.com/v1/query`. It was a success.
4. I even deleted a table entry and then curled again. It was a success and the table was successfully untracked.

Maybe you can try to make the changes manually yourself and see what happens?

**Comment by ElixirMike** [comment]
Ok, I figured it out. In Javascript, when you use the "delete" method to remove a node from a JSON object, it leaves a "null" value. This was causing the issue. I simply needed to use the javascript splice method to remove the node cleanly and then I could run the replace_metadata.

FYI, you must use JSON.Stringify in javascript to convert the JSON object to it's string equivalent for the method to work.

Anyway, all good now.
